src/worker.cc

```diff
@@ -318,6 +318,35 @@ void add_date_to_locations(Options& options,
   }
 }
 
+// Parses JSON rings of the form [[lon1, lat1], [lon2, lat2], ...]] and operates on
+// PBF objects of the sort "repeated Location". Rings can be optionally closed.
+template <typename ring_pbf_t>
+void parse_ring(ring_pbf_t& ring,
+                rapidjson::Value& coord_array,
+                rapidjson::Document::AllocatorType& allocator,
+                bool close_ring = true) {
+  // optionally close the ring
+  if (coord_array[0] != coord_array[coord_array.Size() - 1] && close_ring) {
+    coord_array.PushBack(coord_array[0], allocator);
+  }
+  for (const auto& coords : coord_array.GetArray()) {
+    if (coords.Size() < 2) {
+      throw std::runtime_error("Polygon coordinates must consist of [Lon, Lat] arrays.");
+    }
+
+    double lon = coords[0].GetDouble();
+    lon = midgard::circular_range_clamp<double>(lon, -180, 180);
+    double lat = coords[1].GetDouble();
+    if (lat < -90.0 || lat > 90.0) {
+      throw std::runtime_error("Latitude must be in the range [-90, 90] degrees");
+    }
+
+    auto* ll = ring->add_coords();
+    ll->set_lng(lon);
+    ll->set_lat(lat);
+  }
```

Comment:
```suggestion
  }
  
  // close the ring if it wasnt already
  if (!ring->coords().empty() && *ring->coords().begin() != *ring.coords()->rbegin()) {
    ring->add_coords()->CopyFrom(*ring->coords().begin());
  }
```

Reply:
> i assume boost geometry requires this like most systems?

Actually bg is surprisingly flexible with its inputs validation, it gives you the option if a ring should be closed or not: (https://www.boost.org/doc/libs/1_75_0/libs/geometry/doc/html/geometry/reference/concepts/concept_polygon.html):

We register our custom type with a macro though, I assume it takes the model's default values to do so. Meaning our ring type does expect to be closed I think.

Reply:
Even though it should be possible to compare protobuf messages, it threw an error along the lines of `no conversion to protobuf::StringPiece` coupled with a `no operator==` error, so I resorted to comparing lat & long individually, see below

Reply:
yeah thats another drawback with protobuf. since at the end of the day every type is inherited from the`message` object the most straight forward way to check equality would be serialize both and check the resulting strings are the same. failing that one would need to do some reflection. both of these options sound slow so i bet they opted just not to support it with out maybe an explicit conversion.

Reply:
doesn't sound like a todo here, so I left it as-is